### PR TITLE
Anubhavgoel26/algos

### PR DESCRIPTION
**Pull request by anubhavgoel26**
Added a parallel version of page rank and a (temporary) version of betweenness centrality

### CODE REVIEW

File: include/scea/algo/pr.hpp [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+        galois::iterate(0ul, numNodes),
+        [&](uint64_t i) {
+          g.for_each_edge(i, [&](uint64_t const&) { out_degrees[i]++; });
+        },
+        galois::no_stats(), galois::loopname("ComputeOutDegrees"),
+        galois::steal());
+
+    for (int iter = 0; iter < MAX_ITERATIONS; ++iter) {
+      std::for_each(newRank.begin(), newRank.end(),
+                    [](std::atomic<double>& n) { n.store(0.0); });
```
Comment (meyerzinn) [inline include/scea/algo/pr.hpp:42]: Why `std::for_each` and not `galois::do_all`?

**Review by meyerzinn — CHANGES_REQUESTED** [review]
Let's split page rank and betweenness centrality into separate PRs so that @bozhiyou can implement the rest of betweenness centrality.

File: include/scea/algo/pr.hpp [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+  static void compute(scea::graph::MutableGraph& g) {
+    const uint64_t numNodes = g.size();
+    std::vector<std::atomic<double>> newRank(numNodes);
+    std::vector<double> rank(numNodes, 1.0 / numNodes);
+    std::vector<unsigned> out_degrees(numNodes, 0);
+
+    galois::do_all(
+        galois::iterate(0ul, numNodes),
+        [&](uint64_t i) {
+          g.for_each_edge(i, [&](uint64_t const&) { out_degrees[i]++; });
```
[the remark this answers is not in the document]
Comment (anubhavgoel26) [reply include/scea/algo/pr.hpp:35]: Yes that would be convenient.

**Comment by anubhavgoel26** [comment]
> Let's split page rank and betweenness centrality into separate PRs so that @bozhiyou can implement the rest of betweenness centrality.

Yeah @bozhiyou can you make a separate PR and edit bc.hpp? Meanwhile, this PR already contains plumbing code to bc to edit_scalability so let's merge it in after the changes (worst case - we still have a serial bc implementation)

File: include/scea/algo/pr.hpp (hunk 1 above)
Comment (anubhavgoel26) [reply include/scea/algo/pr.hpp]: Changed.

**Review by meyerzinn — APPROVED** [review]

**Pull request merged**